GPL/Events/Network/Probe.bpf.c

```diff
@@ -43,6 +44,136 @@ static int inet_csk_accept__exit(struct sock *sk)
     return 0;
 }
 
+static int udp_skb_handle(struct sk_buff *skb, enum ebpf_net_udp_info evt_type)
+{
+
+    if (ebpf_events_is_trusted_pid())
+        goto out;
+
+    struct ebpf_dns_event *event = get_event_buffer();
+    if (event == NULL)
+        goto out;
+
+    // read from skbuf
+    unsigned char *skb_head = BPF_CORE_READ(skb, head);
+    // get lengths
+    u16 net_header_offset       = BPF_CORE_READ(skb, network_header);
+    u16 transport_header_offset = BPF_CORE_READ(skb, transport_header);
+
+    u8 proto = 0;
+
+    struct iphdr ip_hdr;
+    bpf_core_read(&ip_hdr, sizeof(struct iphdr), skb_head + net_header_offset);
+    if (ip_hdr.version == 4) {
+        proto = ip_hdr.protocol;
+
+        bpf_probe_read(event->net.saddr, 4, &ip_hdr.saddr);
+        bpf_probe_read(event->net.daddr, 4, &ip_hdr.daddr);
+
+        event->net.family = EBPF_NETWORK_EVENT_AF_INET;
+    } else if (ip_hdr.version == 6) {
+        struct ipv6hdr ip6_hdr;
+        bpf_core_read(&ip6_hdr, sizeof(struct ipv6hdr), skb_head + net_header_offset);
+        proto = ip6_hdr.nexthdr;
+
+        bpf_probe_read(event->net.saddr6, 16, ip6_hdr.saddr.in6_u.u6_addr8);
+        bpf_probe_read(event->net.daddr6, 16, ip6_hdr.daddr.in6_u.u6_addr8);
+
+        event->net.family = EBPF_NETWORK_EVENT_AF_INET6;
+    }
+
+    if (proto != IPPROTO_UDP) {
+        goto out;
+    }
+
+    struct udphdr udp_hdr;
+    bpf_core_read(&udp_hdr, sizeof(struct udphdr), skb_head + transport_header_offset);
+    event->net.dport     = bpf_ntohs(udp_hdr.dest);
+    event->net.sport     = bpf_ntohs(udp_hdr.source);
+    event->net.transport = EBPF_NETWORK_EVENT_TRANSPORT_UDP;
+
+    // filter out non-dns packets
+    if (event->net.sport != 53 && event->net.dport != 53) {
+        goto out;
+    }
+
+    struct task_struct *task = (struct task_struct *)bpf_get_current_task();
+    ebpf_pid_info__fill(&event->pids, task);
+    bpf_get_current_comm(event->comm, TASK_COMM_LEN);
+    event->hdr.ts = bpf_ktime_get_ns();
+
+    // constrain the read size to make the verifier happy
+    // see skb_headlen() in skbuff.h
+    size_t readsize = BPF_CORE_READ(skb, len);
+    size_t datalen  = BPF_CORE_READ(skb, data_len);
+    size_t headlen  = readsize - datalen;
+    // headlen of zero indicates we have no non-paged data, and thus cannot read
+    // anything from the root data node
+    if (headlen == 0) {
+        u32 zero                    = 0;
+        struct ebpf_event_stats *es = bpf_map_lookup_elem(&ringbuf_stats, &zero);
+        if (es != NULL) {
+            es->lost++;
+        }
+        goto out;
+    }
+
+    if (headlen > MAX_DNS_PACKET) {
```

Comment:
```suggestion
    event->orig_len = headlen;
    if (headlen > MAX_DNS_PACKET) {
```

Just so that userland can detect truncation.